### PR TITLE
feat: Support ANY/ALL operators

### PR DESCRIPTION
**Pull request by ovr**
Hello!

Ref issue: https://github.com/sqlparser-rs/sqlparser-rs/issues/476

I want to support `a = ANY(expr)` in DF, but after reviewing I come to decision that it requires changes in `sql-parser` to simplify code (because hacking when the right side is a function is a super tricky)

Thanks

### CODE REVIEW

**Comment by ovr** [comment]
@alamb WDYT? Should I create a separate issue for supporting AllOperatorSubquery & AnyOperatorSubQuery?

**Review by alamb — COMMENTED** [review]
Ho @ovr  --- sounds like an exiting feature!

Looking at the postgres documentation,  https://www.postgresql.org/docs/current/functions-subquery.html it seems to me like `ANY` / `ALL` can be combined with any operator that results in a `boolean`, not just `<, <=, =, !=, >=, >`

For example, with `~`:

```sql
alamb=# select x from test where x ~ ANY (select x from test);
  x  
-----
 foo
(1 row)
```

What would you think about adding something like

```rust
 /// `ANY()`
    Any(Box<Expr>),
```

This might be more in keeping with the spirit of sql-parser-rs to parse the query faithfully without applying any implicit semantics to it. 

> @alamb WDYT? Should I create a separate issue for supporting AllOperatorSubquery & AnyOperatorSubQuery?

I don't think separate issues are necessary

**Comment by alamb** [comment]
> @alamb Oh, I see, but anyway what about LIKE/AND/OR operators which are supported in Binary expression but is not supported for ANY & ALL operators? Another question is, whether it will lead to a nonimplicit pattern machine on the DF side?

I am not quite sure what you mean about "`nonimplicit` pattern match on the DF side"


Maybe you could do something like the following with a match guard (totally untested):

```rust
sqlparser::Expr::BinaryExpr {
  left, 
  op
  right
} if is_any_expr(&right) => {...}
  }
...
```


and 
```rust
/// returns true of `expr` represents an `Any` clause
fn is_any_expr(expr: sqlparser::Expr) ->bool)  {
...
}
```

I found this style worked pretty well in simplify_expressions: 

https://github.com/apache/arrow-datafusion/blob/master/datafusion/core/src/optimizer/simplify_expressions.rs#L498-L508

**Comment by ovr** [comment]
@alamb Oh, I see your idea about simplifying it and wrapping it on the right side of the BinaryExpression expression. I've updated the PR as you suggested. Thanks

File: src/ast/mod.rs [hunk 1]
```diff
@@ -270,6 +270,10 @@ pub enum Expr {
         op: BinaryOperator,
         right: Box<Expr>,
     },
+    /// Any operation e.g. `1 ANY (1)` or `foo > ANY(bar)`, It will be wrapped in the right side of BinaryExpr
+    AnyOp(Box<Expr>),
```
Comment (alamb) [inline src/ast/mod.rs:274]: 👍

File: tests/sqlparser_common.rs [hunk 2]
```diff
@@ -998,6 +998,32 @@ fn parse_bitwise_ops() {
...
 }
 
+#[test]
+fn parse_binary_any() {
+    let select = verified_only_select("SELECT a = ANY(b)");
+    assert_eq!(
+        SelectItem::UnnamedExpr(Expr::BinaryOp {
+            left: Box::new(Expr::Identifier(Ident::new("a"))),
+            op: BinaryOperator::Eq,
+            right: Box::new(Expr::AnyOp(Box::new(Expr::Identifier(Ident::new("b"))))),
```
Comment (alamb) [inline tests/sqlparser_common.rs:1008]: 👍  looks good to me

**Review by alamb — APPROVED** [review]
Looks great to me @ovr  👍

**Pull request merged**